src/lib.rs

```diff
@@ -0,0 +1,221 @@
+                .into_owned(),
+        }
+    }
+}
+
+/// end can be None due to lack of Return callback
+#[cfg(Py_3_8)]
+struct ProfilerEntry {
+    file_name: String,
+    func_name: String,
+    start: Instant,
+    end: Option<Instant>,
+}
+
+/// Profiler context to be used as a value for ContextVar.
+#[pyclass]
+struct ProfilerContext {
+    entries: std::cell::Cell<HashMap<usize, ProfilerEntry>>,
+}
+
+/// Format an entry into file_name, func_name and duration.
+fn format_entry(entry: &ProfilerEntry) -> (String, String, Option<u128>) {
+    let duration = entry
+        .end
+        .and_then(|v| Some(v.duration_since(entry.start).as_nanos()));
+    (entry.file_name.clone(), entry.func_name.clone(), duration)
+}
+
+#[pymethods]
+impl ProfilerContext {
+    #[new]
+    fn new() -> Self {
+        ProfilerContext {
+            entries: std::cell::Cell::new(HashMap::new()),
+        }
+    }
+    #[getter]
+    fn entries(&mut self) -> PyResult<Vec<(String, String, Option<u128>)>> {
+        Ok(self
+            .entries
+            .get_mut()
+            .iter()
+            .map(|(_, entry)| format_entry(entry))
+            .collect())
+    }
+}
+
+fn get_context(py: Python, obj: *mut pyo3::ffi::PyObject) -> Option<Py<ProfilerContext>> {
+    let mut context_obj: *mut pyo3::ffi::PyObject = std::ptr::null_mut();
+
+    unsafe {
+        match PyContextVar_Get(obj, std::ptr::null_mut(), &mut context_obj) {
+            0 => (),
+            _ => return None,
+        };
+    }
+
+    unsafe { Py::from_owned_ptr_or_opt(py, context_obj) }
+}
+
+/// Our profiler callback
+extern "C" fn callback(
+    obj: *mut pyo3::ffi::PyObject,
+    frame: *mut PyFrameObject,
+    what: c_int,
+    _arg: *mut pyo3::ffi::PyObject,
+) -> c_int {
+    let event: TraceEvent = match what.try_into() {
+        Ok(event) => match event {
+            TraceEvent::Call | TraceEvent::Return => event,
+            _ => return 0,
+        },
+        _ => return 0,
+    };
+
+    let frame_data = match FrameData::try_from(frame) {
+        Ok(v) => v,
+        Err(err) => {
+            println!("{}", err);
+            return 0;
+        }
+    };
+
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let context = match get_context(py, obj) {
+        Some(v) => v,
+        _ => return 0,
+    };
+    let mut context = context.as_ref(py).borrow_mut();
+
+    match event {
+        TraceEvent::Call => {
+            let start = Instant::now();
+            let entry = ProfilerEntry {
+                func_name: frame_data.func_name,
+                file_name: frame_data.file_name,
+                start,
+                end: None,
+            };
+            context
+                .entries
+                .get_mut()
+                .insert(frame_data.identifier, entry);
+        }
+        TraceEvent::Return => {
+            if frame_data.is_yielded_coroutine {
+                return 0;
+            }
```

Comment:
!!! There is an issue that pertains to your entire logic. It's the same one about yield that I warned you before. You calculate cumulative time (cumtime) everywhere. Usually you'd want total time (tottime) in asyncio scenarios, as cumtime will include time spent processing other non-related CPU intensive code. !!!

Your code might be perfect per-se, but right now the profiler is unfortunately useless.

You can add a new field for tottime and increase it every yield. It will require that you'll track every call though.